cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -2593,28 +2610,61 @@ int getPodIndexFromPodName(String podName)  {
             }
         }
 
+        /**
+         * Checks if any Kafka broker needs rolling update to add or remove JBOD volumes. If it does, we trigger a
+         * sequential rolling update because the pods need to be rolled in sequence to add or remove volumes.
+         *
+         * @return
+         */
         Future<ReconciliationState> kafkaRollToAddOrRemoveVolumes() {
             Storage storage = kafkaCluster.getStorage();
 
             // If storage is not Jbod storage, we never add or remove volumes
             if (storage instanceof JbodStorage) {
                 JbodStorage jbodStorage = (JbodStorage) storage;
 
-                return kafkaSetOperations.getAsync(namespace, KafkaCluster.kafkaClusterName(name))
-                        .compose(sts -> {
-                            if (sts != null) {
-                                int lastPodIndex = Math.min(kafkaCurrentReplicas, kafkaCluster.getReplicas()) - 1;
-                                return maybeRollKafkaInSequence(sts, pod -> needsRestartBecauseAddedOrRemovedJbodVolumes(pod, jbodStorage, kafkaCurrentReplicas, kafkaCluster.getReplicas()), 0, lastPodIndex);
-                            } else {
-                                // STS does not exist => nothing to roll
-                                return withVoid(Future.succeededFuture());
+                // We first check if any broker actually needs the rolling update. Only if at least one of them needs it,
+                // we trigger it. This check helps to not go through the rolling update if not needed.
+                return podOperations.listAsync(namespace, kafkaCluster.getSelectorLabels())
+                        .compose(pods -> {
+                            for (Pod pod : pods) {
+                                if (!needsRestartBecauseAddedOrRemovedJbodVolumes(pod, jbodStorage, kafkaCurrentReplicas, kafkaCluster.getReplicas()).isEmpty())   {
```

Comment:
Seems like the `not` operator is put on accidentally on this line? This will still call 3 times in case of JBOD config change, is that the expectation?

Reply:
Sorry, but call _what_ 3 times?

Reply:
Sorry, I'm still not sure I follow. If there is any change and the restart is needed, it will call `maybeRollKafkaInSequence` once on line 2639 and with a cluster having a 3 nodes, the recursion will call it two more times. This is correct and expected.

Reply:
ok. thanks. Does each recursion expect to check readiness for all the 3 pods or just pod in question?

Reply:
It checks all pods with the index and earlier. I tried to explain it in the commnents. If it is not clear, please comment on them.